### DISCUSSION TITLE
How to scope selective atoms only?

### DISCUSSION THREAD

**Discussion by jaisharx** [pmndrs/jotai#2154] [Q&A]
I'm having a problem figuring out a way to scope only certain atoms with the current provider methods.

Check out this codesandbox: https://codesandbox.io/s/naughty-rgb-5xn4dg?file=/src/App.tsx

I have a component that I'm rendering multiple times with 3 different atoms. Two of the atoms are supposed to be local, and one is global. I can make the two atoms I want local to each component by wrapping the component with a provider, but in doing so that also makes the global atom local, which I don't want.

Hence I'm wondering if it is possible to only scope certain atoms (or values) with providers or any method?

I'm attaching the code here, you can also checkout the codesandbox above. Thanks for you time already :)
```jsx
import { Provider, atom, useAtom } from "jotai";

const localAtom1 = atom("localAtom1");
const localAtom2 = atom("localAtom2");
const globalAtom = atom("globalAtom");

const TestComp = () => {
  const [globalAtomValue, setGlobalAtomValue] = useAtom(globalAtom);
  const [localAtomValue1, setLocalAtomValue1] = useAtom(localAtom1);
  const [localAtomValue2, setLocalAtomValue2] = useAtom(localAtom2);

  return (
    <div>
      <div>
        {globalAtomValue}
        <input
          value={globalAtomValue}
          onChange={(e) => setGlobalAtomValue(e.target.value)}
        />
      </div>
      <div>
        {localAtomValue1}
        <input
          value={localAtomValue1}
          onChange={(e) => setLocalAtomValue1(e.target.value)}
        />
      </div>
      <div>
        {localAtomValue2}
        <input
          value={localAtomValue2}
          onChange={(e) => setLocalAtomValue2(e.target.value)}
        />
      </div>
    </div>
  );
};

export default function Test() {
  const list = [1, 2, 3];

  return (
    <div style={{ display: "flex" }}>
      {list.map((item) => (
        <Provider key={item}>
          Comp:{item}
          <TestComp key={item} />
        </Provider>
      ))}
    </div>
  );
}
```

**Comment by dai-shi** [comment]
This has been asked several times lately.

Short answer: It's out of the scope of Jotai core library. One third-party solution is `jotai-molecules`, and I would like to look for another lightweight solution looking for a contributor.

Long answer:

Jotai atoms are just like functions.
```js
const localAtom1 = atom("localAtom1");
const localAtom2 = atom("localAtom2");
const globalAtom = atom("globalAtom");
```
At the point of defining atoms, we don't know if an atom is global or local. It just depends on where "store" is located.

Jotai's Provider is simpler wrapper around React Context. It doesn't do anything special. You can check out the source code:
https://github.com/pmndrs/jotai/blob/4d90760ce04c425db6ad1ee89ca8695d2b1d519a/src/react/Provider.ts#L18-L36

Now, if you were to do something you wanted, you'd need to understand "store".
Let's do it without using Jotai's Provider, and just use bare React Context.
We don't usually need it as Jotai's Provider covers this use case, but for learning purpose the bare context would be good.

```js
import { createContext, useContext, useRef } from "jotai";
import { atom, useAtom, getDefaultStore, createStore } from "jotai";

const localAtom1 = atom("localAtom1");
const localAtom2 = atom("localAtom2");
const globalAtom = atom("globalAtom");

const StoreContext = createContext();

const globalStore = getDefaultStore();

const TestComp = () => {
  const localStore = useContext(StoreContext);
  const [globalAtomValue, setGlobalAtomValue] = useAtom(globalAtom, { store: globalStore });
  const [localAtomValue1, setLocalAtomValue1] = useAtom(localAtom1, { store: localStore });
  const [localAtomValue2, setLocalAtomValue2] = useAtom(localAtom2, { store: localStore });

  return (
    <div>
      <div>
        {globalAtomValue}
        <input
          value={globalAtomValue}
          onChange={(e) => setGlobalAtomValue(e.target.value)}
        />
      </div>
      <div>
        {localAtomValue1}
        <input
          value={localAtomValue1}
          onChange={(e) => setLocalAtomValue1(e.target.value)}
        />
      </div>
      <div>
        {localAtomValue2}
        <input
          value={localAtomValue2}
          onChange={(e) => setLocalAtomValue2(e.target.value)}
        />
      </div>
    </div>
  );
};

export default function Test() {
  const list = [1, 2, 3];

  const storeCache = useRef(new Map()).current;
  const getLocalStore = (item) => {
    if (!storeCache.has(item)) {
      storeCache.set(item, createStore());
    }
    return storeCache.get(item)
  }

  return (
    <div style={{ display: "flex" }}>
      {list.map((item) => (
        <StoreContext.Povider key={item} value={getLocalStore(item)}>
          Comp:{item}
          <TestComp key={item} />
        </StoreContext.Provider>
      ))}
    </div>
  );
}
```

The above code is a bit contrived and it can be simplified, but I hope you get the idea.

`jotai-molecules` is a nice library to tackle this kind of use case with _just one store_. It's highly recommended and more preferable than multiple stores.
We would like to come up with a similar idea to support "scope" use case with a single store. Contributor wanted.

**Comment by dai-shi** [comment]
I just developed another util for this kind of use cases. Hope someone can try.
https://github.com/jotaijs/jotai-scope/pull/2
